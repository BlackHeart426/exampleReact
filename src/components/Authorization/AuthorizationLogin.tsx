import React, {ChangeEvent, useEffect, useState} from "react";
import {validateForm} from "../../components/validateForm/validateForm";
import {AnyAction, bindActionCreators, Dispatch} from "redux";
// import {authorizationActionCreator, authorizationGoogleActionCreator} from "../../store/action/authorization";
import {connect} from "react-redux";
import {Form, Button, Container, Col, Row} from "react-bootstrap";
import {TextField} from "@material-ui/core";
import IStoreState from "../../store/IStoreState";
import {signIn} from "../../api/authenticationApi";
import { signInActionCreator } from "../../actions/authentiocation/interfaceAuthentication/authenticationActions";

export interface IState {
    email: {
        status: boolean,
        message: string,
        value: string
    },
    password: {
        status: boolean,
        message: string,
        value: string
    },
    shouldRemember: boolean

}

export interface IAuthorizationLoginProps {
    signIn: (email: string, password: string) => (dispatch: Dispatch<AnyAction>) => Promise<void>
    shouldRemember: boolean;
    onChangeForm: (change: string) => void;
    onHideModal: () => void;
}

export const AuthorizationLogin: React.FC<IAuthorizationLoginProps> = ({
     onChangeForm, onHideModal, shouldRemember
 }) => {

    const initialState: IState = {
        email: {
            status: false,
            message: '',
            value: ''
        },
        password: {
            status: false,
            message: '',
            value: '1'
        },
        shouldRemember: false
    }

    const [state, setState] = useState<IState>(initialState)
    const [password, setPassword] = useState('');
    const [dialogOpened, setDialogOpened] = useState(false);
    const [isButtonDisabled, setIsButtonDisabled] = useState(true);
    const [errorForm, setError] = useState(initialState);

    const handleLogin = () => {
        console.log(state.password.value)
        // props.action.authorization(state.email.value, state.password.value)
        handleClose()
    };

    const handleRememberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { checked } = e.target;
        setState({...state, shouldRemember: checked});
        // props.onRememberChange(checked);
    };


    const handleClose = () => {
        onHideModal()
        setDialogOpened(false)
    };

    useEffect(()=>{
        setState({...state, shouldRemember: shouldRemember})
    },[])

    useEffect(() => {
        if (state.email.value.trim() && state.password.value.trim()) {
            setIsButtonDisabled(false);
        } else {
            setIsButtonDisabled(true);
        }
    }, [state]);

    const handleKeyPress = (e: React.KeyboardEvent) => {
        if (e.keyCode === 13 || e.which === 13) {
            isButtonDisabled || handleLogin();
        }
    };

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>, cb: any) => {
        const {name, value} = e.target;
        const infoValid = validateForm(name, value)
        setError({...errorForm, [name]: {status: infoValid.error, message: infoValid.errorMessage}});
        return cb
    }

    const handl = (e: React.ChangeEvent<HTMLInputElement>) => {
        setPassword(e.target.value)
    }
    return (

        <Form data-testid="login-form">
            <div>{state.password.value}</div>
            <div style={{textAlign: "center"}}>
                <strong>Login</strong>
            </div>
            <TextField
                error={errorForm.email.status}
                helperText={errorForm.email.message}
                variant="outlined"
                fullWidth
                id="email"
                autoFocus
                name="email"
                type="email"
                size={"small"}
                label="Email"
                placeholder="Email"
                margin="normal"
                value={state.email.value}
                onChange={(e: ChangeEvent<HTMLInputElement>) => handleChange(e, setState({...state, email: {value: e.target.value, message: '', status: false}}))}
                onKeyPress={(e)=>handleKeyPress(e)}
            />
            <Form.Group controlId="formLogin">
                <br />
                <Form.Label>Password</Form.Label>
                <Form.Control
                    data-testid="password"
                    name="password"
                    type="password"
                    value={state.password.value}
                    placeholder="Password"
                    // onChange={handl}
                    onChange={(e: ChangeEvent<HTMLInputElement>) => handleChange(e, setState({...state, password: {value: e.target.value, message: '', status: false}}))}
                    // onKeyPress={(e)=>handleKeyPress(e)}
                />
            </Form.Group>
            <Form.Check
                data-testid="remember"
                name="remember"
                type="checkbox"
                onChange={handleRememberChange}
                checked={state.shouldRemember}
                id={`default-checkbox`}
                // onClick={handleChange}
                label={`Remember me`}
            />
            <Button
                type="submit"
                block
                variant={"primary"}
                onClick={handleLogin}
                disabled={isButtonDisabled}
            >
                LOGIN
            </Button>
            <Container>
                <Row  >
                    <a href="#" onClick={() => onChangeForm('recovery')}  >
                        Forgot password?
                    </a>
                    <a href="#" onClick={() => onChangeForm('signUp')}>
                        Don't have an account? Sign Up
                    </a>
                </Row>
            </Container>
        </Form>
    )
}

function mapDispatchToProps(dispatch: Dispatch<AnyAction>) {
    return {
        signIn: bindActionCreators((email, password) => signInActionCreator(email, password), dispatch)
    }
}

export default connect(null, mapDispatchToProps)(AuthorizationLogin)