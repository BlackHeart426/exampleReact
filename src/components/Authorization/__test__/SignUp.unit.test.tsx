import React, {Component, useEffect} from "react";
import {AuthorizationLogin, IProps} from "../AuthorizationLogin"
import { render, fireEvent, waitForElement } from "@testing-library/react";
import {shallow} from "enzyme";


describe('<AuthorizationLogin/>',  () => {
    const mockLogin = jest.fn()
    const password = '4TYE46E8RT!@qw'
    const name = 'register'
    const wrongPassword = '11'
    const email = 'vas@gmail.com'
    const wrongEmail = 'vascom'

    const register = shallow<Component>(<AuthorizationLogin shouldRemember={false} onChangeForm={(name: string) => mockLogin(name)} onHideModal={mockLogin}/>)

    it('renders properly', () => {
        expect(register).toMatchSnapshot()
    })

    it('render firm with disabled button at initial render', ()=>{
        console.log(register.find('Button').debug())
        expect(register.find('Button').prop('disabled')).toBeTruthy()
    })

    it('render firm with disabled button at initial render', ()=>{

        register.find('[type="password"]').simulate('change', {
            target: {
                name: 'password',
                value: password
            }
        })
        register.find('[type="email"]').simulate('change', {
            target: {
                name: 'email',
                value: email
            }
        })
        const useEffect = jest.spyOn(React, "useEffect").mockImplementation(f=>f())
        console.log(register.find('[type="email"]').prop('value'))
        console.log(register.find('[type="password"]').prop('value'))
        expect(register.find('Button').prop('disabled')).toBeFalsy()
    })

    it('should set the password value on change event with trim', () => {
        register.find('[type="password"]').simulate('change', {
            target: {
                name: 'password',
                value: password,
            },
        });
        expect(register.find('[type="password"]').prop('value')).toEqual(
            password,
        );
    });

    it('should set the email value on change event with trim', () => {
        register.find('[type="email"]').simulate('change', {
            target: {
                name: 'email',
                value: email,
            },
        });
        expect(register.find('[type="email"]').prop('value')).toEqual(
            email,
        );
    });
    //
    // it('should set wrong email value', () => {
    //     register.find('[type="email"]').simulate('change', {
    //         target: {
    //             name: 'email',
    //             value: wrongEmail,
    //         },
    //     });
    //     expect(register.find('[type="email"]').prop('value')).toEqual(
    //         email,
    //     );
    // });
    // it('should set wrong password value', () => {
    //     register.find('[type="password"]').simulate('change', {
    //         target: {
    //             name: 'password',
    //             value: wrongPassword,
    //         },
    //     });
    //     expect(register.find('[type="password"]').prop('value')).toEqual(
    //         password,
    //     );
    // });

})