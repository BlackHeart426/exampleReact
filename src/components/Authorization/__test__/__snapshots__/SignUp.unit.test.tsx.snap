// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorizationSignUp/> renders properly 1`] = `
<Fragment>
  <div>
    <WithStyles(ForwardRef(Typography))
      align="center"
    >
      <strong>
        Sign Up
      </strong>
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(TextField))
      className="makeStyles-contentText-2"
      error={false}
      fullWidth={true}
      helperText=""
      id="username"
      label="Username"
      margin="normal"
      name="username"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Username"
      size="small"
      type="text"
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      className="makeStyles-contentText-2"
      error={false}
      fullWidth={true}
      helperText=""
      id="email"
      label="Email"
      margin="normal"
      name="email"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Email"
      size="small"
      type="email"
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      InputProps={
        Object {
          "endAdornment": <ForwardRef(WithStyles)
            position="end"
          >
            <ForwardRef(WithStyles)
              aria-label="toggle password visibility"
              edge="end"
              onClick={[Function]}
              size="small"
            >
              <UNDEFINED />
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>,
        }
      }
      className="makeStyles-contentText-2"
      error={false}
      fullWidth={true}
      helperText=""
      id="password"
      label="Password"
      margin="normal"
      name="password"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Password"
      size="small"
      type="password"
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      InputProps={
        Object {
          "endAdornment": <ForwardRef(WithStyles)
            position="end"
          >
            <ForwardRef(WithStyles)
              aria-label="toggle password visibility"
              edge="end"
              onClick={[Function]}
              size="small"
            >
              <UNDEFINED />
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>,
        }
      }
      className="makeStyles-contentText-2"
      error={false}
      fullWidth={true}
      helperText=""
      id="passwordRepeat"
      label="Password Repeat"
      margin="normal"
      name="passwordRepeat"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Repeat Password"
      size="small"
      style={
        Object {
          "marginTop": "8px",
        }
      }
      type="password"
      variant="outlined"
    />
    <WithStyles(ForwardRef(FormControl))
      className="makeStyles-action-1"
      fullWidth={true}
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={true}
        onClick={[Function]}
        size="large"
        type="button"
        variant="contained"
      >
        Sign Up
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(Grid))
      container={true}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={true}
      >
        <WithStyles(ForwardRef(Link))
          href="#"
          onClick={[Function]}
          variant="body2"
        >
          Forgot password?
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(Link))
          href="#"
          onClick={[Function]}
          variant="body2"
        >
          Login
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </div>
</Fragment>
`;
